File: 0027-remove-element/main.go
package main
import ("fmt")

// beats 100%
func removeElement(nums []int, val int) int {
  count := 0
	for i := 0; i < len(nums); i++ {
		e := nums[i]

		// 0 <= nums[i] <= 50
		// -1 is nil
		if e == val || e == -1 {
			rest := nums[(i + 1):len(nums)]

			nextValIndex := -1
			out:
			for ii := 0; ii < len(rest); ii++ {
				ee := rest[ii]
				if ee != -1 && ee != val {
					nextValIndex = ii
					break out
				}
			}

			if nextValIndex != -1 {
				nextVal := nums[i + nextValIndex + 1]
				nums[i] = nextVal
				nums[i + nextValIndex + 1] = -1 // nil
				count += 1
			} else if i == len(nums) - 1 {
				nums[i] = -1
			} else {
				// next
			}
		} else {
			count += 1
		}
	}
	return count
}

func main () {
	nums := []int{3,2,2,3};
	val := 3;
	r := removeElement(nums, val)
	fmt.Println(r == 2)
}

File: 0027-remove-element/main.js
// @param {number[]} nums
// @param {number} val
// @return {number}

// beats 100%
var removeElement = function(nums, val) {
  let count = 0;
  for (let i = 0; i < nums.length; i++) {
    let e = nums[i];
    if (e === val || e === null) {
      let rest = nums.slice((i + 1), nums.length)
      let j = null;
      for (let ii = 0; ii < rest.length; ii++) {
        let ee = rest[ii];
        if (ee !== null && ee !== val) {
          j = ii;
          break;
        }
      }
      if (j != null) {
        let nextVal = nums[i + j + 1];
        nums[i] = nextVal;
        nums[i + j + 1] = null;
        count += 1;
      } else if (i === nums.length - 1) {
        nums[i] = null;
      } else {
        // next
      }
    } else {
      count += 1;
    }
  }
  return count;
};

// let nums = [0,1,2,2,3,0,4,2];
// let val = 2;
// let r = removeElement(nums, val)
// console.log(r == 5)


let nums = [3,2,2,3];
let val = 3;
let r = removeElement(nums, val)
console.log(r == 2)